rotkehlchen/accounting/structures/evm_event.py

```diff
@@ -0,0 +1,228 @@
+import json
+import logging
+import typing
+from typing import Any, Final, Optional, cast
+from rotkehlchen.accounting.structures.balance import Balance
+
+from rotkehlchen.accounting.structures.base import HistoryBaseEntry
+from rotkehlchen.accounting.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.assets.asset import Asset
+from rotkehlchen.chain.evm.types import string_to_evm_address
+from rotkehlchen.errors.misc import InputError
+from rotkehlchen.errors.serialization import DeserializationError
+from rotkehlchen.fval import FVal
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.serialization.deserialize import deserialize_optional
+from rotkehlchen.types import ChecksumEvmAddress, Location, TimestampMS
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+EVM_EVENT_FIELDS = tuple[
+    Optional[str],  # protocol
+    Optional[str],  # product
+    Optional[str],  # address
+    Optional[str],  # extra_data
+]
+
+EVM_EVENT_FIELDS_NO_EXTRA_DATA = tuple[
+    Optional[str],  # protocol
+    Optional[str],  # product
+    Optional[str],  # address
+]
+
+EVM_EVENT_FIELDS_COUNT = len(typing.get_args(EVM_EVENT_FIELDS))
+
+
+EVM_EVENT_DB_TUPLE_READ = tuple[
+    int,            # identifier
+    bytes,          # event_identifier
+    int,            # sequence_index
+    int,            # timestamp
+    str,            # location
+    Optional[str],  # location label
+    str,            # asset
+    str,            # amount
+    str,            # usd value
+    Optional[str],  # notes
+    str,            # type
+    str,            # subtype
+    str,            # address
+    Optional[str],  # protocol
+    Optional[str],  # product
+    Optional[str],  # extra_data
+]
+
+
+SUB_SWAPS_DETAILS: Final = 'sub_swaps'
+LIQUITY_STAKING_DETAILS: Final = 'liquity_staking'
+
+ALL_DETAILS_KEYS = {
+    SUB_SWAPS_DETAILS,
+    LIQUITY_STAKING_DETAILS,
+}
+
+
+def get_tx_event_type_identifier(event_type: HistoryEventType, event_subtype: HistoryEventSubType, protocol: str) -> str:  # noqa: E501
+    return str(event_type) + '__' + str(event_subtype) + '__' + protocol
+
+
+class EvmEvent(HistoryBaseEntry):
+    def __init__(
+            self,
+            *,  # enforce kwargs for this constructor
```

Comment:
If to allow args -- then how to sort the arguments in the constructor? 

Since `address` is always required it should go before optional fields like `location_label`, `identifier`, etc., which belong to `HistoryBaseEntry`.

And then I think it will be quite easy to make a mistake by using a wrong order of arguments. 

We don't use args for such construstors anyway -- so do you still think that enforcing kwargs is a bad idea here?

Reply:
I believe we never use the arguments without kwargs but I believe is fine to reorder them, they are different classes and if you use kwargs the order doesn't matter, the position argument is used if no kwargs are provided afaik